bd-report-writer/src/ffi.rs

```diff
@@ -0,0 +1,569 @@
+pub struct BDAppMetrics {
+  app_id: *const c_char,
+  version: *const c_char,
+  version_code: i64,
+  cf_bundle_version: *const c_char,
+  running_state: *const c_char,
+  memory_used: u64,
+  memory_free: u64,
+  memory_total: u64,
+}
+
+#[repr(C)]
+pub struct BDDeviceMetrics {
+  time_seconds: u64,
+  time_nanos: u32,
+  timezone: *const c_char,
+  manufacturer: *const c_char,
+  model: *const c_char,
+  os_version: *const c_char,
+  os_brand: *const c_char,
+  os_fingerprint: *const c_char,
+  os_kernversion: *const c_char,
+  power_state: i8,
+  power_charge_percent: u8,
+  network_state: i8,
+  architecture: i8,
+  display_height: u32,
+  display_width: u32,
+  display_density_dpi: u32,
+  platform: i8,
+  rotation: i8,
+  cpu_abi_count: u8,
+  cpu_abis: *const *const c_char,
+}
+
+#[repr(C)]
+pub struct BDBinaryImage {
+  id: *const c_char,
+  path: *const c_char,
+  load_address: u64,
+}
+
+#[repr(C)]
+pub struct BDCPURegister {
+  name: *const c_char,
+  value: u64,
+}
+
+#[repr(C)]
+pub struct BDStackFrame {
+  type_: i8,
+  frame_address: u64,
+  symbol_address: u64,
+  symbol_name: *const c_char,
+  class_name: *const c_char,
+  file_name: *const c_char,
+  line: i64,
+  column: i64,
+  image_id: *const c_char,
+  state_count: usize,
+  state: *const *const c_char,
+  reg_count: usize,
+  regs: *const BDCPURegister,
+}
+
+#[repr(C)]
+pub struct BDThread {
+  name: *const c_char,
+  state: *const c_char,
+  active: bool,
+  index: u32,
+  priority: f32,
+  quality_of_service: i8,
+}
+
+/// Convert a boxed `ReportProcessor` into a raw pointer for use over the
+/// foreign function interface. This conversion avoids running the Box's (and
+/// processor's) destructors.
+///
+/// * `handle_ptr_ptr` - the raw pointer to the processor
+/// * `proc_box`       - boxed `ReportProcessor` instance
+macro_rules! pack_processor {
+  ($handle_ptr_ptr:ident, $proc_box:ident) => {
+    unsafe {
+      *$handle_ptr_ptr = Box::into_raw($proc_box) as *const _;
+    }
+  };
+}
+
+/// Turn a raw pointer into a boxed `ReportProcessor` for use. Should be paired
+/// with `pack_processor!()`.
+///
+/// * `handle_ptr_ptr`    - the raw pointer to the processor
+/// * `failure_ret_value` - value to return upon failure to unpack
+macro_rules! unpack_processor {
+  ($handle_ptr_ptr:ident, $failure_ret_value:expr) => {
+    unsafe {
+      if $handle_ptr_ptr.is_null() || (*$handle_ptr_ptr).is_null() {
+        return $failure_ret_value;
+      }
+      Box::from_raw(*$handle_ptr_ptr as *mut ReportProcessor<'_>)
+    }
+  };
+}
+
+/// Entry point to creating a new Report
+///
+/// * `report_type` - raw value of the `ReportType` which will be written
+/// * `sdk_id`      - identifier of the integration sending the report
+/// * `sdk_version` - version of the integration sending the report
+///
+/// # Safety
+/// This function returns an object which is not managed by normal lifetime
+/// semantics and must be discarded using `bdrw_dispose_buffer_handle()`
```

Comment:
i think pedantically this would point out the expectations of the parameters that make this function safe to call, e.g. that all the pointers are dereferncable and that `handle` can be written to

not sure if you want to care about putting all the unsafe assumptions in the safety block or just make reasonable assumptions about usage